Bootstrap

andela-iikikin opened the pull request:
# Get Peer Review for the basic functionalities of this URL shortening application
## What does this PR do?

- **Authenticates users using both time based token authentication and email authentication**

- [ ] Users can register on the app and use their registration credentials to request tokens

- [ ] Tokens are used for authentication in the API, and emails are only used to request tokens on the API
- **Shortening URL service**
- [ ] Users can send in a long URL and expect to get a shortened version of it.

- [ ] Registered users can send in a long URL and include a vanity string and expect to receive a shortened version composed of the vanity string

- [ ]  Registered users can view statistics such as

1. Number of visits to a particular short URL

2. Information of visitor to a short URL

## How should this be manually tested?

- Instructions have been included in  the README file.

I-chat committed:
[Chore] Add User model tests

I-chat committed:
[Chore]Add ShortUrl model test

I-chat committed:
[Chore]Add LongUrl model test

I-chat committed:
[Feature]Add app models

I-chat committed:
[Chore]Add more tests for app models

I-chat committed:
[Feature]Add time column for ShortUrl

I-chat committed:
[Chore]Change variable name to more descriptive names

I-chat committed:
[Feature]Add support for test coverage

I-chat committed:
[Chore]Remove unused import and correct typos

I-chat committed:
[Chore]Add api authentication tests

I-chat committed:
[Feature]Setup api blueprint

I-chat committed:
[Chore]Add registration test

I-chat committed:
[Feature]Add api auth and registration

I-chat committed:
[Chore]Remove redundant shell context

I-chat committed:
[Chore]Add more tests for sad paths

Change single quotes to double quotes

I-chat committed:
[Feature]Add custom error and input validation

Correct typo errors

I-chat committed:
[Chore]Add requirement and correct typos

Add important packages to requirements.txt and correct typographical errors.
Add coverage html tmp folder to .gitignore

I-chat committed:
[Chore]Arrange imports alphabetically and create simple frontend view

I-chat committed:
[Chore]Test Url shortening

I-chat committed:
[Feature]Add vanity_string input check

Change error messages to be explanatory

I-chat committed:
[Feature]Add visitors log models

Add a UrlActivityLogs class that maps to the visitors log database and establish a many to many relationship with the short_url table

I-chat committed:
[Chore]Move vanity_string check to shorten.py

I-chat committed:
[Chore]Add more test cases

I-chat committed:
[Chore]Change error message to be more explanatory

I-chat committed:
[Chore]Add test for url management functionalities

I-chat committed:
[Feature]Change defualt when value

Change the default when value of ShortUrl model to make use of SQLalchemy timestamps

I-chat committed:
[Chore]Add more test cases to cover sad paths

I-chat committed:
[Chore]Move imports for unittest to cover them

I-chat committed:
[Feature]Add more test and add functionalites

Add functionalities for shortening long_urls, redirecting short_urls, sorting_urls, deactivating and activating short_urls
Add more test cases to cover sad paths

I-chat committed:
[Chore]Correct flake8 errors

I-chat committed:
[Feature]Setup bootstrap login modal

Install flask-bootstrap and added it to requirements.txt.
Create a Sign up modal with bootstrap

I-chat committed:
[Feature]Register app authentication blueprint

I-chat committed:
[Fix]Fix vanity_string with whitespaces

I-chat committed:
[Chore]Move functions and modules to enable reusability

Rename urlshortener.py to helper.py to make it more descriptive and moved some functions to models.py
Move helper.py to app folder to make it easier to import from main

I-chat committed:
[Feature]Implement the landing page view

I-chat committed:
[Feature]Create the shortening input form

Create a new widget to support submit buttons

I-chat committed:
[Feature]Create the registration and login forms

I-chat committed:
[Feature]Create the main app authentication blueprint

I-chat committed:
[Chore]Add README.md file

Add travis integration and coveralls badge
Add scrutinizer and license badge

I-chat committed:
[Fix]Added API version to API URL

Change test to call new API URL

I-chat committed:
[Feature]Add logging of visitors info to db

I-chat committed:
[Chore]Add test for new endpoints

Test for getting a user short_url list
Test the logging of visitors info to db

I-chat committed:
[Feature]Add authentication view functionality

I-chat committed:
[Chore]Setup app for Heroku and Travis deployment

Update requirement.txt file
Setup Profile and runtime.txt for heroku
Set database URl for heroku support
Setup coveralls and scrutinzer badges

[Fix]Correct travis run test command

[Fix]Set database uri and remove unused method

[Fix]Install coveralls for travis

[Fix]Update requirement.txt with latest packages

[Fix]Keep .coverage file for coveralls

[Chore]Add requirements for heroku deployment

[Chore]Edit heroku Procfile

[Fix]Add postgres for heroku deployment

I-chat committed:
[Fix]Fix wrong indentation

I-chat committed:
[Fix]Fix uncaught condition

I-chat committed:
[Fix]Correct wrong status codes

hassanoloye commented on app/api/errors.py:
```diff
@@ -0,0 +1,49 @@
+from flask import request, render_template, jsonify
+
+
+@api.app_errorhandler(404)
```
Love the way you customize and handle errors here

hassanoloye reviewed (commented):
Nice work overall. Code looks clean and almost perfect

I-chat committed:
[Feature]Implement the form for anonymous shortenening

I-chat committed:
[Feature]Implement the landing page views

I-chat committed:
[Chore]Change the file names of templates to style the index page

I-chat committed:
[Fix]Set all routes to strictly False

I-chat committed:
[Fix]Append custom message to 404 defualt message

I-chat committed:
[Fix]Correct error in updating long url many tables

Change error message to more descriptive messages

I-chat committed:
[Fix]Fix change target url functionality

Exclude app auth folder from test

I-chat committed:
[Fix]Correct typos and custom messages

I-chat committed:
[Chore]Add more test to cover sad paths

kosyfrances commented on app/__init__.py:
```diff
@@ -1,17 +1,29 @@
+from flask import Flask
+from flask_bootstrap import Bootstrap
+from flask_sqlalchemy import SQLAlchemy
 from flask_login import LoginManager
```
According to [PEP8 docs](https://www.python.org/dev/peps/pep-0008/#imports), 
Imports should be grouped in the following order:

* standard library imports
* related third party imports
* local application/library specific imports
* You should put a blank line between each group of imports.

In your case, `from config import Config` should be imported last after a blank line from the flask imports.

andela-iikikin replied:
Ok, will be implementing these changes.

kosyfrances commented on app/api/__init__.py:
```diff
@@ -0,0 +1,6 @@
+
+api = Blueprint('api', __name__)
+
+from . import auth, errors, shorten
```
Same comment as above with [imports](https://www.python.org/dev/peps/pep-0008/).

```Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.```

andela-iikikin replied:
I am not sure if this is avoidable as everything I am importing here are defined within the api Blueprint declared above it. Thanks though

kosyfrances commented on app/api/auth.py:
```diff
@@ -0,0 +1,67 @@
+from ..models import User
+from voluptuous import MultipleInvalid
+from flask_httpauth import HTTPBasicAuth
+from .validators import register
```
Same comment as above with [imports](https://www.python.org/dev/peps/pep-0008/).

andela-iikikin replied:
Noted

kosyfrances commented on app/api/errors.py:
```diff
@@ -0,0 +1,50 @@
+from . import api
+from .auth import auth
+from flask import request, render_template, jsonify
+
```
Same comment as above with [imports](https://www.python.org/dev/peps/pep-0008/).

andela-iikikin replied:
Would definitely change this.

kosyfrances commented on app/api/shorten.py:
```diff
@@ -0,0 +1,256 @@
+from ..helper import UrlSaver
+from .validators import valid_url
+from voluptuous import MultipleInvalid
+
```
Same comment as above with [imports](https://www.python.org/dev/peps/pep-0008/).

andela-iikikin replied:
Noted

kosyfrances commented on app/api/shorten.py:
```diff
@@ -0,0 +1,256 @@
+
+    long_url = request.json.get('url')
+    result = short_url.change_long_url(long_url, g.current_user)
+    if result is True:
```
According to [pep 8 programming recommendations](https://www.python.org/dev/peps/pep-0008/#programming-recommendations), 
```Don't compare boolean values to True or False using == .

Yes:   if greeting:
No:    if greeting == True:
Worse: if greeting is True:
```
In this case, `if result:` is a better option to use.

andela-iikikin replied:
This would definitely be implemented.

kayodeadeniyi commented on app/api/auth.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@api.route('/register', methods=['POST'], strict_slashes=False)
+def new_user():
```
@andela-iikikin Is this just a method 😳 ? It's too long

kosyfrances commented on app/api/validators.py:
```diff
@@ -0,0 +1,66 @@
+
+def check_vanity_str(value):
+    """Check that all vanity strings are strings and non empty."""
+    if type(value) is str and len(value) > 0 and " " not in value:
```
Empty strings are faulty in python, so instead of having something like `.... " " not in value`, we can have something like 
```if value and type(value) is str: ``` to check if the value is a string and not an empty one. However, if `len(value) > 0` may not be a good way to check because a string with multiple spaces in between would pass with that check. You probably might want to strip leading and trailing spaces from value to be safe.

kosyfrances commented on app/auth/__init__.py:
```diff
@@ -0,0 +1,5 @@
+
+auth = Blueprint('auth', __name__)
+
+from . import views, auth
```
Check my previous comments about import style in python.

kosyfrances commented on app/auth/forms.py:
```diff
@@ -0,0 +1,31 @@
+from wtforms.validators import Email, Length, Required, Regexp, EqualTo
+from wtforms import ValidationError
+from ..models import User
+
```
Check my previous comments about import style in python.

kosyfrances commented on app/auth/views.py:
```diff
@@ -0,0 +1,60 @@
+from . import auth
+from ..main.forms import ShortForm
+from ..models import User
+from .forms import RegistrationForm, LoginForm
```
Check my previous comments about import style in python.

kosyfrances commented on app/main/views.py:
```diff
@@ -1,7 +1,59 @@
+from ..helper import UrlSaver
+from ..models import User
+from .forms import ShortForm
 
```
Check my previous comments about import style in python.

kosyfrances commented on app/models.py:
```diff
@@ -0,0 +1,232 @@
+from itsdangerous import (TimedJSONWebSignatureSerializer
+                          as Serializer, BadSignature, SignatureExpired)
+from werkzeug.security import generate_password_hash, check_password_hash
+
```
Check my previous comments about import style in python.

kayodeadeniyi commented on app/static/font-awesome-4.7.0/less/path.less:
```diff
@@ -0,0 +1,15 @@
+    url('@{fa-font-path}/fontawesome-webfont.woff?v=@{fa-version}') format('woff'),
+    url('@{fa-font-path}/fontawesome-webfont.ttf?v=@{fa-version}') format('truetype'),
+    url('@{fa-font-path}/fontawesome-webfont.svg?v=@{fa-version}#fontawesomeregular') format('svg');
+  // src: url('@{fa-font-path}/FontAwesome.otf') format('opentype'); // used when developing fonts
```
@andela-iikikin Remove dead code

kayodeadeniyi commented on app/static/font-awesome-4.7.0/scss/_path.scss:
```diff
@@ -0,0 +1,15 @@
+    url('#{$fa-font-path}/fontawesome-webfont.woff?v=#{$fa-version}') format('woff'),
+    url('#{$fa-font-path}/fontawesome-webfont.ttf?v=#{$fa-version}') format('truetype'),
+    url('#{$fa-font-path}/fontawesome-webfont.svg?v=#{$fa-version}#fontawesomeregular') format('svg');
+//  src: url('#{$fa-font-path}/FontAwesome.otf') format('opentype'); // used when developing fonts
```
@andela-iikikin Remove dead codes

njirap replied:
consider using something like a `pre-commit` git hook to prevent this

kosyfrances commented on run.py:
```diff
@@ -1,23 +1,44 @@
+    COV = coverage.coverage(branch=True, include='app/*', omit=['app/main/*', 'app/auth/*'])
+    COV.start()
+
+from app import create_app, db
```
Check my previous comments about import style in python.

kosyfrances commented on run.py:
```diff
@@ -1,23 +1,44 @@
+        import sys
+        os.environ['FLASK_COVERAGE'] = '1'
+        os.execvp(sys.executable, [sys.executable] + sys.argv)
     import unittest
```
Check my previous comments about import style in python.

kosyfrances commented on tests/test_api_auth.py:
```diff
@@ -0,0 +1,90 @@
+from base64 import b64encode
+from flask import url_for
+import json
+
```
Check my previous comments about import style in python.

kosyfrances commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,161 @@
+from app.models import User, ShortUrl, LongUrl
+from datetime import datetime
+from flask_sqlalchemy import sqlalchemy
+from time import sleep
```
Check my previous comments about import style in python.

kayodeadeniyi reviewed (commented):
I was successful at doing just few nitpicks on this PR. Looking great but it's quite massive. I will advise you get some more pythonistas to check it out

njirap commented on app/api/auth.py:
```diff
@@ -0,0 +1,67 @@
+def verify_password(email_or_token, password):
+    """Manage the authentication of either email and password or token."""
+    if not email_or_token:
+        g.current_user = AnonymousUserMixin()
```
instantiating a Mixin.. this is new.

njirap commented on app/api/auth.py:
```diff
@@ -0,0 +1,67 @@
+    if not password == confirm_password:
+        abort(400, "password and confirm_password do not match.")
+
+    if User.query.filter_by(email=email).first() is not None:
```
why not remove `is not None`

osule commented on tests/test_url_shortening.py:
```diff
@@ -0,0 +1,206 @@
+        shorten_response_with_token_auth2 = self.client.post(
+                            url_for('api.shorten_url'), headers=token_header,
+                            data=post_data4)
+        json_shorten_response_with_token = json.loads(
```
you may be able to retrieve the json representation of the response object from any client request call with   `response.json()`. e.g.
```python
json_shorten_response_with_token  = shorten_response_with_token_auth2.json()
```

andela-iikikin replied:
I think it's only Request objects that has a json attribute but `shorten_response_with_token_auth2` here is a Response object. Thanks for the suggestion though

osule commented on tests/test_register.py:
```diff
@@ -0,0 +1,162 @@
+        json_response = json.loads(response.data.decode("utf-8"))
+        msg = json_response["message"]
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(msg, "The request is invalid or inconsistent. Invalid"
```
Multiline strings can be written using python docstrings.

osule commented on app/helper.py:
```diff
@@ -0,0 +1,66 @@
+    @staticmethod
+    def save_url(short_url, long_url, user):
+        """Save and append short and long urls to their various tables."""
+        long_url_exist = LongUrl.query.filter_by(long_url=long_url).first()
```
long_url_exist is misleading as this isn't boolean.

andela-iikikin replied:
Thank you, the name would be changed.

osule commented on app/helper.py:
```diff
@@ -0,0 +1,66 @@
+            user.short_urls.append(shorturl)
+            long_url_exist.users.append(user)
+            db.session.commit()
+            return shorturl
```
We can find common logic with both cases that can be moved right after the condition.

andela-iikikin replied:
Thanks, would implement these changes.

osule commented on app/api/validators.py:
```diff
@@ -0,0 +1,66 @@
+    """Check that value has a valid email pattern."""
+    if email(value):
+        return value
+    abort(400, "Not a valid email format.")
```
raising a validation error with validation functions will be better semantic.

I-chat committed:
[Fix]Fix PEP8 import errors

changed miss leading long_url_exist variable name
remove checking empty string with length of string
remove initializing of Mixin
remove dead codes
raise validation error with validation functions

jubrilissa commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    if not short_url:
+        abort(404, "You are yet to shorten any URL.")
+    return jsonify({'short_url list':
+                    [{'Date_added': x.when, 'Times_visted': x.no_of_visits,
```
You might want to have a specific format for your property names. either snake_case, camelCase, ProperCamelCase. But try to avoid using all

jubrilissa commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+@api.route('/short_url/<int:id>/activate', methods=['PUT'],
+           strict_slashes=False)
+@auth.login_required
+def toogle_is_active(id):
```
Not sure if what you meant here is toggle.

I-chat committed:
[Feature]Add landing html page

I-chat committed:
[Feature]Add anonymous users ajax request to endpoint

I-chat committed:
[Feature]Add copy short_url to clipboard feature

I-chat committed:
[Feature]Add image files

I-chat committed:
[Feature]Add user dashboard with sidebars and dropdowns

erika-dike commented on app/__init__.py:
```diff
@@ -1,17 +1,30 @@
+    app.register_blueprint(auth_blueprint, url_prefix='/auth')
+    app.register_blueprint(api_blueprint, url_prefix='/api/v1')
+    login_manager.init_app(app)
     return app
```
There should be a blank line at the end of your python files.

erika-dike commented on app/api/auth.py:
```diff
@@ -0,0 +1,68 @@
+    user = User.query.filter_by(email=email_or_token).first()
+    if not user:
+        return False
+    g.current_user = user
```
What happens when the password is found to be wrong in the line below this? Shouldn't this be done after user password has been verified?

erika-dike commented on app/api/auth.py:
```diff
@@ -0,0 +1,68 @@
+@auth.login_required
+def get_token():
+    """Return a valid token for registered users."""
+    if g.current_user.is_anonymous or g.token_used:
```
Are you saying that a user can't request a token once he has gotten one?

andela-iikikin replied:
Not with a token. He can with an email and password combination though.

erika-dike commented on app/api/auth.py:
```diff
@@ -0,0 +1,68 @@
+    """Return a valid token for registered users."""
+    if g.current_user.is_anonymous or g.token_used:
+        return abort(403)
+    return jsonify({'token': g.current_user.generate_auth_token(
```
I would advise you get the token before this line or you can break it this way:
```
return jsonify({
    'token': g.current_user.generate_auth_token(
        expiration=3600).decode('ascii'),
    'expiration': '3600secs'
}), 200
```

andela-iikikin replied:
Thanks. Will do this.

erika-dike commented on app/api/auth.py:
```diff
@@ -0,0 +1,68 @@
+        abort(400, "password and confirm_password do not match.")
+
+    if User.query.filter_by(email=email).first():
+        abort(400, "User already registered.Try login.")  # existing user
```
No space before Try.

erika-dike replied:
I think more information will be nice here like `Someone has already registered with that email address. Try logging in.`

andela-iikikin replied:
Flake8 linters does flag an error for blank lines after a function docstring. I will be definitely be implementing your second suggestion though.

erika-dike commented on app/api/errors.py:
```diff
@@ -0,0 +1,51 @@
+
+@api.app_errorhandler(400)
+def bad_request(e):
+    """Customize the defualt http 400 status_code message."""
```
typo: default

andela-iikikin replied:
Corrected thanks

erika-dike commented on app/api/errors.py:
```diff
@@ -0,0 +1,51 @@
+
+@api.errorhandler(403)
+def forbidden(e):
+    """Customize the defualt http 403 status_code message."""
```
typo @ default

erika-dike commented on app/api/errors.py:
```diff
@@ -0,0 +1,51 @@
+        'error': '401 Unauthenticated',
+        'message': 'The authentication credentials sent with the request are '
+                   'invalid or insufficient for the request.'
+        })
```
Nitpick: this is how to close a multiline statement when you start the first parameter or item on a new line:
```
response = jsonify({
    'error': '401 Unauthenticated',
    'message': 'The authentication credentials sent with the request are '
               'invalid or insufficient for the request.'
})
```

andela-iikikin replied:
Noted

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+from app import db
+from app.api import api
+from app.api.validators import valid_url
+from app.api.auth import auth
```
You forgot to place this in alphabetical order.

andela-iikikin replied:
Thanks for the reminder

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+
+def check_authentication_with_token():
+    """Check that a user is authenticated and used token for authentication."""
+    if g.current_user.is_anonymous or not g.token_used:
```
Won't `g.token_used` cover the first condition as well?

andela-iikikin replied:
No. `g.token_used` only checks if the user used a token for authentication or not. The user could possibly have used an email and password combination for authentication.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+
+    long_url = request.json.get('url')
+
+    if request.json.get('vanity_string') and not g.current_user.is_anonymous:
```
I thought anonymous users won't get here as per line 37.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    user = g.current_user
+
+    short_url = UrlSaver.generate_and_save_urls(long_url, user)
+    return jsonify({'id': short_url.id,
```
I think line 52 and 58 can be consolidated. You could even stretch by consolidating 50 and 57. For example by passing a None value as vanity string for unauthenticated users.

andela-iikikin replied:
Sure, will do.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    elif short_url and short_url.is_active:
+        return redirect(short_url.long_url.long_url, code=302)
+    else:
+        abort(404)
```
Could this be refactored to avoid the repetition of short_url in every conditional statement? I think so :).

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+        abort(400, "The request does not contain a body.")
+
+    try:
+        short_url = [short_url for short_url in g.current_user.short_urls
```
This is a potentially long operation as a user creates more short urls. Why not use the more performant database searching capabilities to get a faster search?

andela-iikikin replied:
Sure, will change this

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    long_url = request.json.get('url')
+    result = short_url.change_long_url(long_url, g.current_user)
+    if isinstance(result, ShortUrl):
+        short_url = site_url + result.short_url
```
Could the site_url be gotten from the request?

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+        abort(404, "You do not have any such URL.")
+
+    long_url = request.json.get('url')
+    result = short_url.change_long_url(long_url, g.current_user)
```
I tend to avoid having a method return two different types. Some painful bugs could come out of that. Instead, raise an exception.

erika-dike replied:
or return true or false

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    """Return a list of a users service shortened URLs."""
+    check_authentication_with_token()
+
+    short_url = [short_url for short_url in g.current_user.short_urls]
```
**Notes:**

+ I think short_urls would be a better name here.
+ Won't something like: short_urls = g.current_user.short_urls.all(), be sufficient here. It should return an enumerable which you should then be able to run a for loop on.

Using the db for operations within its capacity is preferrable to performing same using your programming language.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+
+    short_url = [short_url for short_url in g.current_user.short_urls]
+    if not short_url:
+        abort(404, "You are yet to shorten any URL.")
```
I suggest "You are yet to shorten a URL" or "you haven't shortened any URLs"

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+
+    try:
+        logs = [short_url for short_url in g.current_user.short_urls
+                if short_url.id == id][0].logs
```
Again, this is a good candidate for the database. A database would perform this in a faster running time than your current O(n) operation.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    if logs:
+        return jsonify({'short_url logs':
+                        [{'I.P Address': x.ip, 'User agent': x.browser,
+                          'System platform': x.platform} for x in logs]}), 200
```
Consider separating placing each key on a separate line to facilitate easy reading. You might go on to consider doing this outside this jsonify block.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+        short_url.is_active = False
+        output = "Deactivation Successful."
+    elif not short_url.is_active and request.url.endswith('activate') and not\
+            request.url.endswith('deactivate'):
```
I wonder if this last condition is necessary. Do you envisage a situation where a URL that leads to this function would simultaneously end with `activate` and `deactivate`.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+        short_url.is_active = True
+        output = "Activation Successful."
+    else:
+        abort(400, "You are trying to either activate a URL that is already"
```
One thing you could do here rather than print both for activation and deactivation is sth like:
`You are trying to {operation} a URL that is already {operation}d.format(operation=request.url.split('/')[-1])`. 
Alternatively, you could get operation type earlier in the function and save it in a variable.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    check_authentication_with_token()
+    users = User.query.all()
+    sorted_users = sorted(users, key=lambda user: len(user.short_urls),
+                          reverse=True)
```
Again, it is preferrable here to have the database return this sorted than perform it here.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+@api.route('/<string:url_type>/<string:sort_type>', strict_slashes=False)
+@auth.login_required
+def sort_urls(url_type, sort_type):
+    """Return short_urls based popularity and recently added."""
```
Did you mean `based on`?

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+        return jsonify({'url_list': longurl_list}), 200
+    elif sort_type == 'date' and url_type == 'shorturl':
+        shorturl_list = ShortUrl.sort_short_url_by_date_added()
+        return jsonify({'url_list': shorturl_list}), 200
```
Nitpick: I think the three return statements are similar and could be sent once at the end of this function.

erika-dike commented on app/api/shorten.py:
```diff
@@ -0,0 +1,257 @@
+    if short_url.deleted:
+        abort(404, "This URL has been deleted.")
+    short_url.deleted = True
+    db.session.commit()
```
Though this is fine, you could consider having a delete method to keep model code out of view code.

erika-dike commented on app/api/validators.py:
```diff
@@ -0,0 +1,21 @@
+                        Schema, Url)
+
+register = Schema({
+    'firstname': Match('^[A-Za-z][A-Za-z]*$', msg="Invalid name format. Name"
```
I think you should make these error message more specific for your users e.g. First name should.....

erika-dike replied:
Nitpick: I am curious as to why you use the current regex expression rather than this shorter version ^[A-Za-z]+$.

erika-dike commented on app/api/validators.py:
```diff
@@ -0,0 +1,21 @@
+    'email': Email(msg="Not a valid email format."),
+    'password': All(str, Length(min=1), msg="Invalid password format."
+                    " Passwords can only be strings."),
+    'confirm_password': All(str, Length(min=1), msg="Invalid password format."
```
Do you really accept a password with a minimum of one character? If so, repent now :)

erika-dike commented on app/auth/forms.py:
```diff
@@ -0,0 +1,32 @@
+    submit = SubmitField('Sign up')
+
+    def validate_email(self, field):
+        if User.query.filter_by(email=field.data).first():
```
Consider using the sqlalchemy exists() construct which returns true or false which is what you are really checking for here.

erika-dike commented on app/auth/views.py:
```diff
@@ -0,0 +1,60 @@
+    if register_form.errors:
+        flash([value for value in register_form.errors.values()][0][0],
+              'register')
+    return render_template('index.html', register_form=register_form,
```
Is there a reason why you aren't separating explicitly POST and GET part of this function and the one below?

erika-dike commented on app/auth/views.py:
```diff
@@ -0,0 +1,60 @@
+    short_form = ShortForm()
+    if login_form.validate_on_submit():
+        user = User.query.filter_by(email=login_form.email.data).first()
+        if user is not None and user.verify_password(login_form.password.data):
```
More pythonic to say `if user and ....`. I probably saw this kind of usage before but forgot to flag it. So if you accept this suggestion, you may want to check for other occurrences.

erika-dike commented on app/auth/views.py:
```diff
@@ -0,0 +1,60 @@
+def logout():
+    logout_user()
+    flash('You have been logged out.')
+    return redirect(url_for('main.index'))
```
Just because I am seeing this a lot, I will say it again. Python files should end with a blank line. You should be using a linter with your editor that warns you of this kind of style violations.

andela-iikikin replied:
I think GitHub denotes a lack of trailing new line with the red [no-newline octicon](https://octicons.github.com/icon/no-newline/). I had a blank line at the end of my files
![dqbls](https://cloud.githubusercontent.com/assets/25608911/25784740/202bdb3e-336a-11e7-93ed-b980570358db.png)

erika-dike commented on app/helper.py:
```diff
@@ -0,0 +1,59 @@
+    @staticmethod
+    def generate_and_save_urls(url, user, vanity_string=None):
+        """Help manage the generating and saving of urls."""
+        if url in [x.long_url for x in user.long_urls]:
```
I think you can query the db for this.

erika-dike commented on app/helper.py:
```diff
@@ -0,0 +1,59 @@
+        if url in [x.long_url for x in user.long_urls]:
+            long_url = LongUrl.query.filter_by(long_url=url).first()
+            return [short_url for short_url in
+                    user.short_urls if short_url.long_url_id == long_url.id][0]
```
Again, another db operation.

erika-dike commented on app/helper.py:
```diff
@@ -0,0 +1,59 @@
+
+        elif vanity_string:
+            short_url = ShortUrl.query.filter_by(
+                                    short_url=vanity_string).first()
```
This line isn't properly formatted. Again, consider installing a linter enforcing PEP8.

erika-dike commented on app/helper.py:
```diff
@@ -0,0 +1,59 @@
+            return [short_url for short_url in
+                    user.short_urls if short_url.long_url_id == long_url.id][0]
+
+        elif vanity_string:
```
I am not sure how this works but is it possible that a match on the first condition also comes with a vanity string?

erika-dike commented on app/helper.py:
```diff
@@ -0,0 +1,59 @@
+            return save_data
+
+    @staticmethod
+    def save_url(short_url, url, user):
```
This method should me in the models. In fact, you could refactor this so that most if not all the methods here become methods in the ShortUrl model.

Favour fat models that contain methods and properties that describe their attributes and behaviour without a need to travel outside.

erika-dike commented on app/helper.py:
```diff
@@ -0,0 +1,59 @@
+        """Generate a unique short_url."""
+        short_url = ''.join(random.choice(
+            string.ascii_letters + string.digits) for _ in range(6))
+        while ShortUrl.query.filter_by(short_url=short_url).first():
```
Consider exists() for this condition

erika-dike commented on app/helper.py:
```diff
@@ -0,0 +1,59 @@
+        while ShortUrl.query.filter_by(short_url=short_url).first():
+            short_url = ''.join(random.choice(
+                string.ascii_letters + string.digits) for _ in range(6))
+        return(short_url)
```
Why the change of style here?

erika-dike commented on app/main/views.py:
```diff
@@ -1,7 +1,59 @@
+    if short_form.errors:
+        flash([value for value in short_form.errors.values()][0][0],
+              'shorten')
+    try:
```
What does this block do? Is it necessary?

andela-iikikin replied:
Whenever a form fails validation there exist a `.errors` property which is a dict object with the fields that failed validation as keys and a list of error messages as the values. I was trying to get a list of all the error messages from the dict to display to the user. As you would know the `.values()` property of a dictionary returns a dict_values object which does not support indexing, so I had to use a list comprehension to convert it to a python list before using the index to access the error message to flash.

erika-dike commented on app/main/views.py:
```diff
@@ -1,7 +1,59 @@
+@main.route('/dashboard')
+@login_required
+def user_dashboard():
+    return render_template('dashboard.html', register_form="",
```
Why add the forms as parameter? It doesn't seem like you need the at all.

andela-iikikin replied:
I see now that there are better ways to handle template variables that could be none if not passed in certain views. This will be removed.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+        except SignatureExpired:
+            return None  # valid token, but expired
+        except BadSignature:
+            return None  # invalid token
```
While your comments tell us both cases are different, your code handles both the same way.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+
+
+class AnonymousUser(AnonymousUserMixin):
+    """Create an User with with AnonymousUserMixin properties."""
```
double with

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+    is_active = db.Column(db.Boolean, default=True)
+    no_of_visits = db.Column(db.Integer, default=0)
+    deleted = db.Column(db.Boolean(), default=False, index=True)
+    when = db.Column(db.DateTime, default=db.func.current_timestamp(),
```
Nitpick: while `when` is fine, there are standard ways of representing this field such as date_created, created_timestamp e.t.c.

andela-iikikin replied:
Thanks, I will update this.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+    user = db.relationship("User", back_populates="short_urls")
+    long_url_id = db.Column(db.Integer, db.ForeignKey('long_url.id'),
+                            nullable=False)
+    long_url = db.relationship("LongUrl", back_populates="short_urls")
```
I don't think you need to have both user_id and user in this model since user_id can be accessed from the user object. Same goes for long_url. http://docs.sqlalchemy.org/en/latest/orm/basic_relationships.html#one-to-many

andela-iikikin replied:
My understanding is that the `user` property needs the foreign key column to be defined in other to return any user object associated with the short_url. I think it's the `user` property that is expendable here but that will mean writing more code to query any associated user using the foreign key column. I am yet to see any associative table without a foreign key column.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+            db.session.delete(old_long_url)
+        self.long_url = new_long_url
+        new_long_url.users.append(user)
+        self.when = db.func.current_timestamp()
```
In this case, I think your when indicates date_modified. You see you need a more explicit and recognizable name. Moreover, you don't need to update this yourself. The db fields have options that can handle this for you. Checkout default and onupdate.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+        self.long_url = new_long_url
+        new_long_url.users.append(user)
+        self.when = db.func.current_timestamp()
+        self.no_of_visits = 0
```
This should also be set as a default.

erika-dike replied:
However, if you are modifying this as I suspect now that you are since this could be an already existing long_url, I wonder why you are resetting it.

andela-iikikin replied:
I am trying to reset it because I believe both short_url and long_url point to the same resource. So if long_url has been visited `n` times, so should short_url and if the user changes the long_url of a particular short_url and I don't reset it, that truth won't be reflected.

erika-dike replied:
Oh I see. Thanks.

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+        active and not deleted.
+        """
+        short_url = ShortUrl.query.filter_by(short_url=short_url).first()
+        if short_url and not short_url.deleted and short_url.is_active:
```
Can a short url be both active and deleted?

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+                'short_url': x.short_url} for x in short_urls]
+
+    @staticmethod
+    def findurl(short_url, request):
```
pythonic: find_url

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+
+    __tablename__ = 'long_url'
+    id = db.Column(db.Integer, primary_key=True)
+    long_url = db.Column(db.String, unique=True)
```
This comment applies both to shortUrl. Why have a model named LongUrl and then have a field named long_url? Why not name the field url?

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+    def sort_long_url_by_date_added():
+        """Sort the long_urls based on how recently they were added.
+
+        Sorts the list of long_urls queried from the LongtUrl
```
typo: LongUrl

erika-dike commented on app/models.py:
```diff
@@ -0,0 +1,244 @@
+        return "<LongUrl(long_url='%s')>" % self.long_url
+
+
+class UrlActivityLogs(db.Model):
```
Consider adding a timestamp field to this model. I haven't seen logs without them.

erika-dike commented on config.py:
```diff
@@ -12,23 +12,19 @@ class Config:
     SQLALCHEMY_COMMIT_ON_TEARDOWN = True
     SITE_URL = dotenv.get('SITE_URL')
```
Since you have SITE_URL in config, why do you still get the one from the environment in your code?

andela-iikikin replied:
I actually did not have it in config at first but later decided it was better to have there and retrieve it from the config whenever I need it rather than from the environment but I can see I forgot to change the rest of my code to reflect this. I will do that right away, thank you

erika-dike commented on tests/test_api_auth.py:
```diff
@@ -0,0 +1,92 @@
+
+    def test_verify_password_with_token(self):
+        """Test verify_password function with valid token."""
+        self.user.save()
```
Why didn't you complete this transaction in the setup?

andela-iikikin replied:
Since I didn't need the user in the database for all the test case, I felt it would be better to save the user in test cases where they are needed but I can see I have only one test case in which the user is not needed in the database so I can just as well ignore it that one time.

erika-dike commented on tests/test_api_auth.py:
```diff
@@ -0,0 +1,92 @@
+        json_response = json.loads(response.data.decode('utf-8'))
+        token = json_response['token']
+        self.assertTrue(token)
+        self.assertEqual(response.status_code, 201)
```
I would advise that you compare status code before response data. This is so you have a good guess where to start your debugging if there are errors.

erika-dike commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,163 @@
+        """Test that same passwords generates different hash values."""
+        self.user.password = '123456'
+        self.user2.password = '123456'
+        self.assertTrue(self.user.password_hash != self.user2.password_hash)
```
This test while okay may not be necessary since you are just testing external code.

erika-dike commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,163 @@
+        self.assertTrue(self.user.password_hash != self.user2.password_hash)
+
+    def test_saving_null_to_users_table(self):
+        """Test that that Users table raises errors for empty columns."""
```
Both test name and comment are misleading. I see that you are instead testing for duplicate users.

erika-dike commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,163 @@
+        self.assertEqual(self.user.verify_auth_token(user_token), self.user)
+        sleep(2)
+        self.assertIsNone(self.user.verify_auth_token(user_token))
+        self.assertIsNone(self.
```
This test is testing several properties of the token. Consider breaking into simpler methods.

erika-dike commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,163 @@
+        self.long_url.users.append(self.user)
+        db.session.add(self.short_url)
+        db.session.commit()
+        self.assertIs(type(self.short_url.when), datetime)
```
Why not use assertInstance(..., datetime)?

andela-iikikin replied:
IsInstance should work, I will change this.

erika-dike commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,163 @@
+        with self.assertRaises(sqlalchemy.exc.IntegrityError):
+            db.session.commit()
+
+    def test_shorturl_is_active_defualts(self):
```
typo on defaults

erika-dike commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,163 @@
+        db.session.commit()
+        self.assertTrue(self.short_url.is_active)
+        self.short_url.is_active = False
+        self.assertFalse(self.short_url.is_active)
```
I don't think the last two lines are necessary for this test.

andela-iikikin replied:
True, I will get rid of it.

erika-dike commented on tests/test_app_models.py:
```diff
@@ -0,0 +1,163 @@
+        self.long_url2.short_urls.append(self.short_url)
+        self.long_url2.users.append(self.user)
+        db.session.add_all([self.user, self.short_url, self.short_url2,
+                            self.long_url, self.long_url2])
```
So I have tried to ignore this but I see that it is cumbersome and possibly unnecessary. If a user has a long URL, this long URL could have several short URLs. A user is related to a long URL and a long URL is related to short URL. My thinking is if I have the appended a short URL to the long URL then I should have access to the short URL from the user model or am I wrong?

erika-dike commented on tests/test_register.py:
```diff
@@ -0,0 +1,162 @@
+        """Delete app and db instances after each test case."""
+        db.session.remove()
+        db.drop_all()
+        self.app_context.pop()
```
Since you perform this app creation in (I think) all your tests. You might consider having a Base class that gets this setup for other tests to inherit and consume.

erika-dike commented on tests/test_url_shortening.py:
```diff
@@ -0,0 +1,206 @@
+                (token + ':' + '').encode('utf-8')).decode('utf-8'),
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
```
The comment about Base class would also include these methods or you could have a testing_utils file to house these.

erika-dike commented on tests/test_url_shortening.py:
```diff
@@ -0,0 +1,206 @@
+        shorten_response_with_email_auth = self.client.post(
+            url_for('api.shorten_url'), headers=email_header,
+            data=self.post_data)
+        shorten_response_with_token_auth = self.client.post(
```
Since this is the actual endpoint you want to test, focus on it and avoid calling the other ones above. You could still obtain the token by calling the function that generates it directly as you did in the user model test.

andela-iikikin replied:
Noted

erika-dike commented on tests/test_url_shortening.py:
```diff
@@ -0,0 +1,206 @@
+                         ' access to this resource.')
+        self.assertIn('short_url', json_shorten_response_with_token)
+        self.assertEqual(403, shorten_response_with_email_auth.status_code)
+        self.assertEqual(201, shorten_response_with_token_auth.status_code)
```
This method is a bit difficult to read. Consider revising.

andela-iikikin replied:
While it should be possible for registered users to access the api endpoints with tokens, their email and password combination should be rejected. That was what I was trying to test. I would be breaking this method into two and then get rid of the long names to give it clarity. Thanks though

erika-dike commented on tests/test_url_shortening.py:
```diff
@@ -0,0 +1,206 @@
+        json_token_response = json.loads(token_response.data.decode('utf-8'))
+        token = json_token_response['token']
+        token_header = self.api_token_auth_headers(token)
+        shorten_response_with_token_auth = self.client.post(
```
Again aim for one response per method, and you would have no need for these long names you are currently using.

andela-iikikin replied:
Thanks, these changes would definitely be implemented

erika-dike commented on tests/test_url_shortening.py:
```diff
@@ -0,0 +1,206 @@
+        token_header = self.api_token_auth_headers(token)
+        shorten_response_with_token_auth = self.client.post(
+                            url_for('api.shorten_url'), headers=token_header,
+                            data=self.post_data3)
```
PEP8 format style violation!!!

andela-iikikin replied:
I would change this. Thank you

erika-dike commented on tests/test_url_shortening.py:
```diff
@@ -0,0 +1,206 @@
+                            url_for('api.shorten_url'), headers=token_header,
+                            data=self.post_data3)
+        json_shorten_response_with_token = json.loads(
+            shorten_response_with_token_auth.data.decode('utf-8'))
```
Sometimes, when you want to facilitate reading of lines like this, consider:
```
json_shorten_response_with_token = json.loads(
    shorten_response_with_token_auth.data.decode('utf-8')
)
```

andela-iikikin replied:
Will fix this. Thank you.

erika-dike reviewed (commented):
Awesome work man. I am sorry for the late response. See my comments inline.

I-chat committed:
[Chore]Implement peer review feedback

Adhere to PEP8 styling
Correct typos and get site_url from config
Break large test and reduce request calls per test
